### PR TITLE
Fix memory leaks #6256

### LINKED ISSUE #6256: Fix memory leaks

### PR DESCRIPTION
**Pull request by oa-s**
Fixes #6256

### CODE REVIEW

**Comment by oa-s** [comment]
should be before #6255

File: modules/AlphaWalletFoundation/AlphaWalletFoundation/Tokens/Helpers/JsonFromTokenUri.swift [hunk 1]
```diff
@@ -14,38 +14,47 @@ import Combine
...
     func fetchJsonFromTokenUri(forTokenId tokenId: String, tokenType: TokenType, address: AlphaWallet.Address, enjinToken: GetEnjinTokenQuery.Data.EnjinToken?) -> AnyPublisher<NonFungibleBalanceAndItsSource<JsonString>, SessionTaskError> {
         return Just(tokenId)
             .receive(on: queue)
             .setFailureType(to: SessionTaskError.self)
-            .flatMap { [queue, getTokenUri] tokenId -> AnyPublisher<NonFungibleBalanceAndItsSource<JsonString>, SessionTaskError> in
+            .flatMap { [weak self, queue, weak getTokenUri] tokenId -> AnyPublisher<NonFungibleBalanceAndItsSource<JsonString>, SessionTaskError> in
+                guard let strongSelf = self, let getTokenUri = getTokenUri else { return .fail(.cancelledError) }
                 let key = "\(tokenId).\(address.eip55String).\(tokenType.rawValue)"
 
-                if let promise = self.inFlightPromises[key] {
```
Comment (hboon) [inline modules/AlphaWalletFoundation/AlphaWalletFoundation/Tokens/Helpers/JsonFromTokenUri.swift:37]: Maybe not the best way to have accessed `self` previously, but `.flatMap()`'s `transform` arg is escaping, why would it retain `self` and cause a leak?

File: modules/AlphaWalletFoundation/AlphaWalletFoundation/Tokens/Helpers/NonFungibleErc1155JsonBalanceFetcher.swift [hunk 2]
```diff
@@ -20,27 +20,39 @@ class NonFungibleErc1155JsonBalanceFetcher {
...
+    }
+
     func fetchErc1155NonFungibleJsons(enjinTokens: EnjinTokenIdsToSemiFungibles) -> AnyPublisher<[AlphaWallet.Address: [NonFungibleBalanceAndItsSource<NonFungibleFromTokenUri>]], SessionTaskError> {
         return erc1155TokenIdsFetcher
             .detectContractsAndTokenIds()
             .mapError { SessionTaskError(error: $0) }
-            .flatMap { contractsAndTokenIds -> AnyPublisher<Erc1155TokenIds.ContractsAndTokenIds, SessionTaskError> in
-                self.addUnknownErc1155ContractsToDatabase(contractsAndTokenIds: contractsAndTokenIds.tokens)
-            }.flatMap { contractsAndTokenIds -> AnyPublisher<(contractsAndTokenIds: Erc1155TokenIds.ContractsAndTokenIds, tokenIdMetaDatas: [TokenIdMetaData]), SessionTaskError> in
-                self._fetchErc1155NonFungibleJsons(contractsAndTokenIds: contractsAndTokenIds, enjinTokens: enjinTokens)
```
Comment (hboon) [inline modules/AlphaWalletFoundation/AlphaWalletFoundation/Tokens/Helpers/NonFungibleErc1155JsonBalanceFetcher.swift:43]: Does the strong `self` reference here only cause a problem when the upstream code (`detectContractsAndTokenIds()` etc) don't emit any results/error?

File: modules/AlphaWalletFoundation/AlphaWalletFoundation/Tokens/Helpers/JsonFromTokenUri.swift (hunk 1 above)
Comment (oa-s) [reply modules/AlphaWalletFoundation/AlphaWalletFoundation/Tokens/Helpers/JsonFromTokenUri.swift:37]: but this sscaping closure captures self, and self captures publisher, that causes issue. we got strong ref to JsonFromTokenUri, in balacne fetcher, when we delete balance fetcher, e.g: switch rpc server, balance fetcher delates, (that is fine) but with it we want to delete all inner objects. so JsonFromTokenUri stays in memory

File: modules/AlphaWalletFoundation/AlphaWalletFoundation/Tokens/Helpers/NonFungibleErc1155JsonBalanceFetcher.swift (hunk 2 above)
Comment (oa-s) [reply modules/AlphaWalletFoundation/AlphaWalletFoundation/Tokens/Helpers/NonFungibleErc1155JsonBalanceFetcher.swift:43]: it causes an issue when its in progress, we ran function and when in same time switch servers, object should receive cancel as its parent dellocated

File: modules/AlphaWalletFoundation/AlphaWalletFoundation/Tokens/Helpers/JsonFromTokenUri.swift (hunk 1 above)
Comment (oa-s) [reply modules/AlphaWalletFoundation/AlphaWalletFoundation/Tokens/Helpers/JsonFromTokenUri.swift:37]: flatMap from combine creates a new publisher. this escaping block captures self as well as self captures publisher

Comment (hboon) [reply modules/AlphaWalletFoundation/AlphaWalletFoundation/Tokens/Helpers/JsonFromTokenUri.swift:37]: 😭 of course! I had read "escaping" as "non-escaping" (so I thought it should retain `self`). All good!

**Review by hboon — APPROVED** [review]

**Pull request merged**